File: app/assets/javascripts/game/pile/DragToConstraints.js
define(['game/Rank'], function(Rank) {
  return {
    "never": function() {
      return false;
    },

    "empty": function() {
      return this.cards.length === 0;
    },

    "klondike": function(pile) {
      if(this.cards.length === 0) {
        return pile.dragCards[0].rank == Rank.king;
      } else {
        var topCard = this.cards[this.cards.length - 1];
        var firstDraggedCard = pile.dragCards[0];
        if(topCard.suit.color == firstDraggedCard.suit.color) {
          return false;
        } else if(topCard.rank === Rank.ace || firstDraggedCard.rank === Rank.king) {
          return false;
        } else {
          return topCard.rank.value == firstDraggedCard.rank.value + 1;
        }
      }
    },

    "foundation": function(pile) {
      if(pile.dragCards.length === 1) {
        if(this.cards.length === 0) {
          return pile.dragCards[0].rank === Rank.ace;
        } else {
          var topCard = this.cards[this.cards.length - 1];
          var dragCard = pile.dragCards[0];
          return topCard.suit === dragCard.suit && ((topCard.rank === Rank.ace && dragCard.rank === Rank.two) || topCard.rank.value + 1 == dragCard.rank.value);
        }
      } else {
        return false;
      }
    },

    "same-rank": function(pile) {
      if(this.cards.length === 0) {
        return false;
      } else {
        var topCardRank = this.cards[this.cards.length - 1].rank.value;
        var draggedCardRank = pile.dragCards[0].rank.value;
        return topCardRank === draggedCardRank;
      }
    },

    "lower-rank": function(pile) {
      var draggedCardRank = pile.dragCards[0].rank.value;
      if(this.cards.length === 0) {
        return false;
      } else {
        var topCardRank = this.cards[this.cards.length - 1].rank.value;
        if(topCardRank == 2) {
          return draggedCardRank === 14;
        } else {
          console.log(topCardRank + ":" + draggedCardRank);
          return topCardRank === draggedCardRank + 1;
        }
      }
    },

    "alternating-rank": function(pile) {
      if(pile.dragCards.length !== 1) {
        return false;
      } else {
        var topCardRank = this.cards[this.cards.length - 1].rank.value;
        var draggedCardRank = pile.dragCards[0].rank.value;
        if(topCardRank == 13 /* King */) {
          return topCardRank === draggedCardRank + 1;
        } else if(topCardRank == 14 /* Ace */) {
          return draggedCardRank === 2;
        } else if(topCardRank == 2 /* Two */) {
          return draggedCardRank == 14 || draggedCardRank === 3;
        } else {
          return topCardRank === draggedCardRank + 1 || topCardRank === draggedCardRank - 1;
        }
      }
    }
  };
});

File: app/models/game/pile/constraints/DragToConstraints.scala
package models.game.pile.constraints

import models.game.Rank._
import models.game.{Rank, Card}
import models.game.pile.{PileOption, Pile}

case class DragToConstraint(override val id: String, f: (Pile, Seq[Card]) => Boolean) extends PileOption(id)

object DragToConstraints {
  val never = DragToConstraint("never", (pile, cards) => false)

  val empty = DragToConstraint("empty", (pile, cards) => pile.cards.isEmpty)

  val foundationDefault = DragToConstraint("foundation", (pile, cards) => if(cards.length == 1) {
    if(pile.cards.isEmpty && cards(0).r == Rank.Ace) {
      true
    } else if(pile.cards.last.s == cards(0).s && pile.cards.last.r == Rank.Ace && cards(0).r == Rank.Two) {
      true
    } else if(pile.cards.last.s == cards(0).s && pile.cards.last.r.value + 1 == cards(0).r.value) {
      true
    } else {
      false
    }
  } else {
    false
  })

  val klondike = DragToConstraint("klondike", (pile, cards) => if(pile.cards.isEmpty) {
    cards.head.r ==  Rank.King
  } else {
    val topCard = pile.cards.last
    val firstDraggedCard = cards.head
    if(!topCard.u) {
      false
    } else {
      if(topCard.s.color == firstDraggedCard.s.color) {
        false
      } else if(topCard.r == Rank.Ace || firstDraggedCard.r == Rank.King) {
        false
      } else {
        topCard.r.value == firstDraggedCard.r.value + 1
      }
    }
  })

  val sameRank = DragToConstraint("same-rank", (pile, cards) => {
    val topCardRank = pile.cards.last.r
    val firstDraggedCardRank = cards.head.r
    topCardRank == firstDraggedCardRank
  })

  val lowerRank = DragToConstraint("lower-rank", (pile, cards) => {
    val firstDraggedCardRank = cards.head.r
    pile.cards.lastOption match {
      case Some(c) => c.r match {
        case Two => firstDraggedCardRank == Ace
        case _ => c.r.value == firstDraggedCardRank.value + 1
      }
      case None =>  false
    }
  })

  val alternatingRank = DragToConstraint("alternating-rank", (pile, cards) => {
    val topCardRank = pile.cards.last.r
    val firstDraggedCardRank = cards.head.r
    topCardRank match {
      case King => firstDraggedCardRank == Queen
      case Ace => firstDraggedCardRank == Two
      case Two => firstDraggedCardRank == Ace || firstDraggedCardRank == Three
      case _ => topCardRank.value == firstDraggedCardRank.value + 1 || topCardRank.value == firstDraggedCardRank.value - 1
    }
  })
}
